README.md
# Dyframe
> Dynamically render responsive HTML into iframe.

[![Bower Version][bower-image]][bower-url]
[![npm Version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]
[![Coverage Status][coveralls-image]][coveralls-url]

[![Dyframe](demo/demo.png)](http://htanjo.github.io/dyframe/)

[**See demo**](http://htanjo.github.io/dyframe/)

## Getting started
### Install

Download and include `dyframe.js` to your HTML.
Available on [Bower](http://bower.io/) and [npm](https://www.npmjs.com/).

* Download: [**dyframe.js**][uncompressed-url] / [minified][minified-url]
* Bower: `$ bower install dyframe --save`
* npm: `$ npm install dyframe --save`

### Example
```html
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Dyframe</title>
    <style>
      #dyframe {
        width: 360px;
        height: 480px;
        border: 1px solid #ddd;
      }
    </style>
  </head>
  <body>
    <div id="dyframe"></div>
    <script src="dyframe.js"></script>
    <script>
      var element = document.getElementById('dyframe');
      new Dyframe(element, {
        html: '<html><body>Hello, world!</body></html>'
      });
    </script>
  </body>
</html>
```

## Constructor
```js
new Dyframe(element, [options]);
```

- `element` : Target DOM element.
- `options` : Options for rendering HTML content.

### Options

#### html
Type: `String`  
Default: `''`

HTML to render.  
Set whole HTML code including `doctype`, `<html>`, `<head>` and `<body>` tag.

#### width
Type `Number`  
Default: `980` (px)

Width for HTML rendering.  
But if you have `profile` option, `width` value will be ignored. (See below)

#### deviceWidth
Type `Number` | `null`  
Default: `null` (px)

Device width for HTML rendering.  
The HTML scaling can be emulated based on `<meta name="viewport">` when you set number.  
But if you have `profile` option, `deviceWidth` value will be ignored. (See below)

#### profile
Type: `String` | `null`  
Default: `null`

Profile name for device emulation.  
When you set proper profile, the scaling will be emulated using profile setting instead of `width` and `deviceWidth` option.

You can use the following profiles, or create custom profile using `Dyframe.addProfile()`.

- **smartphone:**  
  width: 980, deviceWidth: 375. Same as iPhone 6 portrait.
- **tablet:**  
  width: 980, deviceWidth: 768. Same as iPad Air 2 portrait.

**Tip:** Profiled element has additional class `df-profile-<name>`.
It can be helpful for styling with CSS.

#### interval
Type: `Number`  
Default: `0` (ms)

Interval to skip rendering.  
Frequent re-rendering, such as [live HTML preview](http://htanjo.github.io/dyframe/), could put heavy load on CPU.
To prevent that, you can limit the frequency using this option.

When you set `500` to this option, the actual DOM rendering takes place only once in 500 ms even if `.render()` method called many times.

## Methods
Create "dyframe" object before using methods.

```js
var dyframe = new Dyframe(element, options);
```

### .render([options])
Re-render the preview content.  
If you call this method with argument, the options will be overriden and re-render.

```js
var element = document.getElementById('dyframe');
var dyframe = new Dyframe(element, {
  html: '<html><body>Hello, world!</body></html>'
});

setTiemout(funciton () {

  // Update HTML content
  dyframe.render({
    html: '<html><body>Updated!</body></html>'
  });

}, 1000);
```

### .destroy()
Clean up the target element.

## Customizing

### Dyframe.addProfile(name, profile)
Add custom device profile to the Dyframe global config.

#### name
Type: `String`

Custom profile name.

#### profile
Type: `Object`

Custom profile data.
Need to define `width` and `deviceWidth` property.

```js
// Add custom profile
Dyframe.addProfile('nexus-6', {
  width: 980,
  deviceWidth: 412
});

// Render using "nexus-6" profile
new Dyframe(element, {
  html: '<html><body>Hello, world!</body></html>',
  profile: 'nexus-6'
});

```

## Compatibility

### Browser support
Dyframe works on most modern browsers including smart devices.  
[Tested](https://saucelabs.com/u/dyframe) on the following browsers:

- Internet Explorer (9+)
- Chrome
- Firefox
- Safari
- Opera
- iOS Safari
- Android Browser

### Module interface
- CommonJS
- AMD

## License
Copyright (c) 2015 Hiroyuki Tanjo. Licensed under the [MIT License](LICENSE).

[bower-image]: https://img.shields.io/bower/v/dyframe.svg
[bower-url]: http://bower.io/
[npm-image]: https://img.shields.io/npm/v/dyframe.svg
[npm-url]: https://www.npmjs.com/package/dyframe
[travis-image]: https://img.shields.io/travis/htanjo/dyframe/master.svg
[travis-url]: https://travis-ci.org/htanjo/dyframe
[coveralls-image]: https://img.shields.io/coveralls/htanjo/dyframe/master.svg
[coveralls-url]: https://coveralls.io/r/htanjo/dyframe
[uncompressed-url]: https://github.com/htanjo/dyframe/raw/v0.5.1/dyframe.js
[minified-url]: https://github.com/htanjo/dyframe/raw/v0.5.1/dyframe.min.js


CHANGELOG.md
# Changelog

## 0.5.1 (2015-09-24)
- Implement [UMD](https://github.com/umdjs/umd) (Universal Module Definition) correctly.

## 0.5.0 (2015-04-24)
- Add `interval` option to prevent frequent re-rendering.

## 0.4.0 (2015-04-23)
- Add `.destroy()` method to clean up.
- Support `initial-scale` value of `<meta name="viewport">`.
- Improve tests using [Sauce Labs](https://saucelabs.com/u/dyframe) and [Coveralls](https://coveralls.io/r/htanjo/dyframe).

## 0.3.0 (2015-04-20)
- Add `deviceWidth` option.
- Support custom profile.
- Add class name to element for styling with css easily.
- Change default `profile` value to `null`.

## 0.2.0 (2015-04-16)
- Support AMD.
- Minor bug fix.

## 0.1.0 (2015-04-15)
- Initial release.


test/spec/dyframe-test.js
/* jshint mocha: true, expr: true */
/* global expect */
(function () {
  'use strict';

  var Dyframe = window.Dyframe;
  var element = document.createElement('div');
  var dyframe;

  // Helpers
  var hasClass = function (element, className) {
    if (element.classList) {
      return element.classList.contains(className);
    }
    else {
      return new RegExp('(^|\\s)' + className + '(?!\\S)', 'g').test(element.className);
    }
  };
  var addClass = function (element, className) {
    if (element.classList) {
      element.classList.add(className);
    }
    else {
      element.className += ' ' + className;
    }
  };

  // Set up fixtures
  element.style.width = '100px';
  element.style.height = '200px';
  document.body.appendChild(element);

  // Specs
  describe('Dyframe', function () {

    describe('.addProfile()', function () {

      afterEach(function () {
        dyframe.destroy();
      });

      it('adds active profile for dyframe objects', function () {
        Dyframe.addProfile('custom', {});
        dyframe = new Dyframe(element, {
          profile: 'custom'
        });
        expect(dyframe.hasActiveProfile()).to.be.true;
      });

    });

    describe('Constructor', function () {

      afterEach(function () {
        dyframe.destroy();
      });

      it('creates <div> and <iframe> in the target element', function () {
        dyframe = new Dyframe(element);
        expect(element.querySelector('div')).not.to.be.null;
        expect(element.querySelector('iframe')).not.to.be.null;
      });

      it('creates HTML content in <iframe>', function () {
        dyframe = new Dyframe(element, {
          html: '<html><body>Hello, world!</body></html>'
        });
        var iframe = element.querySelector('iframe');
        var body = iframe.contentWindow.document.body;
        expect(body.innerHTML).to.equal('Hello, world!');
      });

      it('sets default options when no options given', function () {
        var defaults = {
          html: '',
          width: 980,
          deviceWidth: null,
          profile: null,
          interval: 0
        };
        dyframe = new Dyframe(element);
        expect(dyframe.options).to.deep.equal(defaults);
      });

      it('overrides default options when options given', function () {
        var options = {
          html: '<html><body>Hello, Dyframe!</body></html>',
          deviceWidth: 360
        };
        var expected = {
          html: '<html><body>Hello, Dyframe!</body></html>',
          width: 980,
          deviceWidth: 360,
          profile: null,
          interval: 0
        };
        dyframe = new Dyframe(element, options);
        expect(dyframe.options).to.deep.equal(expected);
      });

      it('adds "df-element" class to the target element', function () {
        dyframe = new Dyframe(element);
        expect(hasClass(element, 'df-element')).to.be.true;
      });

      it('adds "df-profile-<name>" class when profile option given', function () {
        dyframe = new Dyframe(element, {
          profile: 'smartphone'
        });
        expect(hasClass(element, 'df-profile-smartphone')).to.be.true;
      });

    });

  });

  describe('dyframe', function () {

    describe('.render()', function () {

      var iframe;
      var body;

      beforeEach(function () {
        dyframe = new Dyframe(element, {
          html: '<html><body>Hello, world!</body></html>'
        });
        iframe = element.querySelector('iframe');
      });

      afterEach(function () {
        dyframe.destroy();
      });

      it('overrides options when argument given', function () {
        dyframe.render({
          width: 1200
        });
        expect(dyframe.options.width).to.equal(1200);
      });

      // Somehow this is not working on PhantomJS...
      // it('makes <iframe> same size with target element', function () {
      //   dyframe.render();
      //   var rect = iframe.getBoundingClientRect();
      //   expect(rect.width).to.equal(100);
      //   expect(rect.height).to.equal(200);
      // });

      it('renders HTML according to default options', function (done) {
        dyframe.render();
        setTimeout(function () {
          expect(iframe.contentWindow.innerWidth).to.equal(980);
          done();
        }, 0);
      });

      it('renders HTML according to width option', function (done) {
        dyframe.render({
          width: 1200
        });
        setTimeout(function () {
          expect(iframe.contentWindow.innerWidth).to.equal(1200);
          done();
        }, 0);
      });

      it('renders HTML accroding to deviceWidth option if HTML has meta-viewport', function (done) {
        dyframe.render({
          html: '<html><head><meta name="viewport" content="width=device-width"></head></html>',
          width: 980,
          deviceWidth: 360
        });
        setTimeout(function () {
          expect(iframe.contentWindow.innerWidth).to.equal(360);
          done();
        }, 0);
      });

      it('renders HTML accroding to width option if HTML does not have meta-viewport', function (done) {
        dyframe.render({
          html: '<html><head></head></html>',
          width: 980,
          deviceWidth: 360
        });
        setTimeout(function () {
          expect(iframe.contentWindow.innerWidth).to.equal(980);
          done();
        }, 0);
      });

      it('renders HTML accroding to width value in meta-viewport', function (done) {
        dyframe.render({
          html: '<html><head><meta name="viewport" content="width=720"></head></html>',
          width: 980,
          deviceWidth: 360
        });
        setTimeout(function () {
          expect(iframe.contentWindow.innerWidth).to.equal(720);
          done();
        }, 0);
      });

      it('renders HTML accroding to initial-scale value in meta-viewport', function (done) {
        dyframe.render({
          html: '<html><head><meta name="viewport" content="initial-scale=0.5"></head></html>',
          width: 980,
          deviceWidth: 360
        });
        setTimeout(function () {
          expect(iframe.contentWindow.innerWidth).to.equal(720);
          done();
        }, 0);
      });

      it('renders HTML ignoring meta-viewport that does not have content', function (done) {
        dyframe.render({
          html: '<html><head><meta name="viewport"></head></html>',
          width: 980,
          deviceWidth: 360
        });
        setTimeout(function () {
          expect(iframe.contentWindow.innerWidth).to.equal(980);
          done();
        }, 0);
      });

      it('renders HTML ignoring meta-viewport that has invalid content', function (done) {
        dyframe.render({
          html: '<html><head><meta name="viewport" content="width:device-width"></head></html>',
          width: 980,
          deviceWidth: 360
        });
        setTimeout(function () {
          expect(iframe.contentWindow.innerWidth).to.equal(980);
          done();
        }, 0);
      });

      it('renders HTML accroding to "smartphone" profile', function (done) {
        dyframe.render({
          html: '<html><head><meta name="viewport" content="width=device-width"></head></html>',
          profile: 'smartphone'
        });
        setTimeout(function () {
          expect(iframe.contentWindow.innerWidth).to.equal(375);
          done();
        }, 0);
      });

      it('renders HTML accroding to "tablet" profile', function (done) {
        dyframe.render({
          html: '<html><head><meta name="viewport" content="width=device-width"></head></html>',
          profile: 'tablet'
        });
        setTimeout(function () {
          expect(iframe.contentWindow.innerWidth).to.equal(768);
          done();
        }, 0);
      });

      it('renders HTML accroding to custom profile', function (done) {
        Dyframe.addProfile('nexus-6', {
          width: 980,
          deviceWidth: 412
        });
        dyframe.render({
          html: '<html><head><meta name="viewport" content="width=device-width"></head></html>',
          profile: 'nexus-6'
        });
        setTimeout(function () {
          expect(iframe.contentWindow.innerWidth).to.equal(412);
          done();
        }, 0);
      });

      it('skips and postpones rendering when interval option given', function (done) {
        dyframe.render({
          html: '<html><body>First rendering</body></html>',
          interval: 1000
        });
        setTimeout(function () {
          body = iframe.contentWindow.document.body;
          expect(body.innerHTML).to.equal('First rendering');
          dyframe.render({
            html: '<html><body>Second rendering</body></html>'
          });
          setTimeout(function () {
            body = iframe.contentWindow.document.body;
            expect(body.innerHTML).to.equal('First rendering');
            setTimeout(function () {
              body = iframe.contentWindow.document.body;
              expect(body.innerHTML).to.equal('Second rendering');
              done();
            }, 1000);
          }, 0);
        }, 0);
      });

    });

    describe('.destroy()', function () {

      beforeEach(function () {
        dyframe = new Dyframe(element, {
          html: '<html><body>Hello, world!</body></html>'
        });
      });

      it('cleans up the target element', function () {
        dyframe.destroy();
        expect(element.innerHTML).to.be.empty;
      });

      it('removes "df" related classes', function () {
        dyframe.render({
          profile: 'smartphone'
        });
        dyframe.destroy();
        expect(hasClass(element, 'df-element')).to.be.false;
        expect(hasClass(element, 'df-profile-smartphone')).to.be.false;
      });

      it('preserves non-"df" classes', function () {
        addClass(element, 'non-df-class');
        dyframe.destroy();
        expect(hasClass(element, 'df-element')).to.be.false;
        expect(hasClass(element, 'non-df-class')).to.be.true;
      });

      it('can be called twice, but does nothing', function () {
        dyframe.destroy();
        expect(element.innerHTML).to.be.empty;
        expect(function () {dyframe.destroy();}).to.not.throw(Error);
        expect(element.innerHTML).to.be.empty;
      });

    });

  });

}());


test/spec/test-dyframe.js
(function () {
  'use strict';

  describe('Dyframe', function () {

    var Dyframe = window.Dyframe;
    var element;
    var dyframe;

    beforeEach(function () {
      element = document.getElementById('dyframe');
      dyframe = new Dyframe(element);
    });

    describe('constructor', function () {

      it('creates <div> and <iframe> in the target element', function () {
        assert.isObject(dyframe.element);
        assert.isNotNull(dyframe.element.querySelector('div'));
        assert.isNotNull(dyframe.element.querySelector('iframe'));
      });

      it('add "df-element" class to the target element', function () {
        assert(dyframe.element.classList.contains('df-element'));
      });

    });

  });

}());


gulpfile.js
'use strict';

var gulp = require('gulp');
var $ = require('gulp-load-plugins')();
var fs = require('fs');
var Karma = require('karma').Server;
var bs = require('browser-sync').create();
var runSequence = require('run-sequence');
var minimist = require('minimist');
var del = require('del');
var opn = require('opn');

var getJson = function (filepath) {
  return JSON.parse(fs.readFileSync(filepath, 'utf8'));
};
var knownOptions = {
  string: 'bump',
  default: {
    bump: 'patch'
  }
};
var options = minimist(process.argv.slice(2), knownOptions);
var banner = [
  '/*!',
  ' * Dyframe',
  ' * @version <%= pkg.version %>',
  ' * @link <%= pkg.homepage %>',
  ' * @author <%= pkg.author %>',
  ' * @license <%= pkg.license %>',
  ' */',
  ''].join('\n');
var scripts = [
  'gulpfile.js',
  'karma.conf.js',
  'src/*.js',
  'demo/*.js',
  'test/spec/*.js'
];

gulp.task('clean', del.bind(null, ['coverage']));

gulp.task('jshint', function () {
  return gulp.src(scripts)
    .pipe($.jshint())
    .pipe($.jshint.reporter('jshint-stylish'))
    .pipe($.if(!bs.active, $.jshint.reporter('fail')));
});

gulp.task('karma', ['clean'], function (callback) {
  new Karma({
    configFile: __dirname + '/karma.conf.js'
  }, callback).start();
});

gulp.task('scripts', function () {
  var pkg = getJson('package.json');
  return gulp.src('src/*.js')
    .pipe($.header(banner, {pkg: pkg}))
    .pipe(gulp.dest('.'))
    .pipe($.uglify({preserveComments: 'some'}))
    .pipe($.rename({suffix: '.min'}))
    .pipe(gulp.dest('.'));
});

gulp.task('serve', ['clean'], function (callback) {
  bs.init({
    server: {
      baseDir: ['src', 'demo'],
    },
    notify: false,
    open: false
  }, function () {
    opn('http://localhost:3000/demo.html');
    callback();
  });
  new Karma({
    configFile: __dirname + '/karma.conf.js',
    singleRun: false
  }).start();
  gulp.watch([
    'demo/*',
    'src/*.js'
  ]).on('change', bs.reload);
  gulp.watch(scripts, ['jshint']);
});

gulp.task('link', function () {
  var downloadBase = 'https://github.com/htanjo/dyframe/raw/';
  var pattern = new RegExp(downloadBase + '.*?[\/]', 'g');
  var version = getJson('package.json').version;
  var downloadDir = downloadBase + 'v' + version + '/';
  return gulp.src(['README.md', 'demo/index.html'], {base: '.'})
    .pipe($.replace(pattern, downloadDir))
    .pipe(gulp.dest('.'));
});

gulp.task('bump', function () {
  return gulp.src(['package.json', 'bower.json'])
    .pipe($.bump({type: options.bump}))
    .pipe(gulp.dest('.'));
});

gulp.task('commit', ['build', 'link'], function () {
  var version = getJson('package.json').version;
  return gulp.src(['package.json', 'bower.json', '*.js', 'README.md', 'demo/index.html'])
    .pipe($.git.commit('Release v' + version));
});

gulp.task('tag', function (callback) {
  var version = getJson('package.json').version;
  $.git.tag('v' + version, 'Release v' + version, callback);
});

gulp.task('deploy', ['test'], function () {
  return gulp.src(['src/*.js', 'demo/**/*', '!demo/demo.html'])
    .pipe($.ghPages());
});

gulp.task('test', ['jshint', 'karma']);

gulp.task('build', ['scripts']);

gulp.task('release', function (callback) {
  runSequence('test', 'bump', 'commit', 'tag', callback);
});

gulp.task('default', function (callback) {
  runSequence('test', 'build', callback);
});


demo/demo.js
(function () {
  'use strict';

  var Dyframe = window.Dyframe;

  // Add custom profile
  Dyframe.addProfile('pc', {
    width: 1366,
    deviceWidth: null
  });
  Dyframe.addProfile('custom', {
    width: 980,
    deviceWidth: 280
  });

  var input = document.getElementById('input');
  var html = input.value;
  var dyframes = [];

  // Base
  var baseElement = document.getElementById('dyframe-base');
  if (baseElement) {
    dyframes.push(new Dyframe(baseElement, {
      html: html,
      interval: 500
    }));
  }

  // Width 1200px
  var widthElement = document.getElementById('dyframe-width');
  if (widthElement) {
    dyframes.push(new Dyframe(widthElement, {
      html: html,
      width: 1200,
      interval: 500
    }));
  }

  // Device width 600px
  var deviceWidthElement = document.getElementById('dyframe-device-width');
  if (deviceWidthElement) {
    dyframes.push(new Dyframe(deviceWidthElement, {
      html: html,
      deviceWidth: 600,
      interval: 500
    }));
  }

  // PC
  var pcElement = document.getElementById('dyframe-pc');
  if (pcElement) {
    dyframes.push(new Dyframe(pcElement, {
      html: html,
      profile: 'pc',
      interval: 500
    }));
  }

  // Tablet
  var tabletElement = document.getElementById('dyframe-tablet');
  if (tabletElement) {
    dyframes.push(new Dyframe(tabletElement, {
      html: html,
      profile: 'tablet',
      interval: 500
    }));
  }

  // Smartphone
  var smartphoneElement = document.getElementById('dyframe-smartphone');
  if (smartphoneElement) {
    dyframes.push(new Dyframe(smartphoneElement, {
      html: html,
      profile: 'smartphone',
      interval: 500
    }));
  }

  // Custom profile
  var customProfileElement = document.getElementById('dyframe-custom-profile');
  if (customProfileElement) {
    dyframes.push(new Dyframe(customProfileElement, {
      html: html,
      profile: 'custom',
      interval: 500
    }));
  }

  // Re-rendering
  var renderElement = document.getElementById('dyframe-render');
  if (renderElement) {
    var renderDyframe = new Dyframe(renderElement, {
      html: html,
      interval: 500
    });
    dyframes.push(renderDyframe);
    var renderProfile = null;
    setInterval(function () {
      switch (renderProfile) {
        case null:
          renderProfile = 'tablet';
          break;
        case 'tablet':
          renderProfile = 'smartphone';
          break;
        case 'smartphone':
          renderProfile = null;
          break;
      }
      renderDyframe.render({
        profile: renderProfile
      });
    }, 3000);
  }

  // Sync HTML content with textarea value
  var updateHtml = function () {
    html = input.value;
    dyframes.forEach(function (dyframe) {
      dyframe.render({
        html: html
      });
    });
  };
  input.addEventListener('change', function () {
    updateHtml();
  });
  input.addEventListener('keydown', function () {
    setTimeout(function () {
      if (input.value === html) {
        return;
      }
      updateHtml();
    }, 0);
  });

  // Hotfix for image rendering
  setTimeout(function () {
    updateHtml();
  }, 1000);

}).call(this);


karma.conf.js
/* jshint camelcase: false */
'use strict';

module.exports = function (config) {

  config.set({
    basePath: '',
    frameworks: ['mocha', 'chai'],
    files: [
      'src/*.js',
      'test/spec/*.js'
    ],
    exclude: [],
    preprocessors: {
      'src/*.js': ['coverage']
    },
    reporters: ['mocha', 'coverage'],
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: true,
    browsers: ['Chrome', 'PhantomJS'],
    singleRun: true,
    client: {
      mocha: {
        reporter: 'html'
      }
    },
    coverageReporter: {
      type : 'lcov',
      dir : 'coverage/'
    },

    // Browser test on BrowserStack
    browserStack: {
      project: 'htanjo/dyframe'
    },
    customLaunchers: {
      'BS_Chrome': {
        base: 'BrowserStack',
        browser: 'chrome',
        os: 'Windows',
        os_version: '8.1'
      },
      'BS_Firefox': {
        base: 'BrowserStack',
        browser: 'firefox',
        os: 'Windows',
        os_version: '8.1'
      },
      'BS_Opera': {
        base: 'BrowserStack',
        browser: 'opera',
        os: 'Windows',
        os_version: '8.1'
      },
      'BS_IE_11': {
        base: 'BrowserStack',
        browser: 'ie',
        browser_version: '11.0',
        os: 'Windows',
        os_version: '8.1'
      },
      'BS_IE_10': {
        base: 'BrowserStack',
        browser: 'ie',
        browser_version: '10.0',
        os: 'Windows',
        os_version: '7'
      },
      'BS_IE_9': {
        base: 'BrowserStack',
        browser: 'ie',
        browser_version: '9.0',
        os: 'Windows',
        os_version: '7'
      },
      'BS_Safari': {
        base: 'BrowserStack',
        browser: 'safari',
        os: 'OS X',
        os_version: 'Yosemite'
      },
      'BS_iOS': {
        base: 'BrowserStack',
        device: 'iPhone 5S',
        os: 'ios',
        os_version: '7.0'
      },
      'BS_Android': {
        base: 'BrowserStack',
        device: 'Samsung Galaxy S5',
        os: 'android',
        os_version: '4.4'
      }
    }
  });

  // Override concfig for CI environment
  if (process.env.CI) {
    if (process.env.BROWSER_STACK_USERNAME && process.env.BROWSER_STACK_ACCESS_KEY) {
      config.browsers = Object.keys(config.customLaunchers);
    }
    else {
      config.browsers = ['PhantomJS'];
    }
    config.reporters.push('coveralls');
  }

};


dyframe.js
(function (global, factory) {
  'use strict';

  if (typeof define === 'function' && define.amd) {
    define(factory);
  }
  else if (typeof module === 'object' && module.exports) {
    module.exports = factory();
  }
  else {
    global.Dyframe = factory();
  }

}(this, function () {
  'use strict';

  // Device profiles
  var profiles = {
    // iPhone 6 portrait
    smartphone: {
      width: 980,
      deviceWidth: 375
    },
    // iPad Air 2 portrait
    tablet: {
      width: 980,
      deviceWidth: 768
    }
  };

  // Prefix for class names
  var prefix = 'df-';

  // Default options
  var defaults = {
    html: '',
    width: 980,
    deviceWidth: null,
    profile: null,
    interval: 0
  };

  // Constructor
  var Dyframe = function (element, options) {
    this.element = element;
    this.wrapper = document.createElement('div');
    this.viewport = document.createElement('iframe');
    this.width = 0;
    this.height = 0;
    this.queued = false;
    this.waiting = false;
    addClass(this.element, prefix + 'element');
    setStyles(this.wrapper, {
      position: 'relative',
      display: 'block',
      height: 0,
      padding: 0,
      overflow: 'hidden'
    });
    setStyles(this.viewport, {
      position: 'absolute',
      top: 0,
      left: 0,
      bottom: 0,
      height: '100%',
      width: '100%',
      border: 0,
      webkitTransformOrigin: '0 0',
      msTransformOrigin: '0 0',
      transformOrigin: '0 0'
    });
    this.wrapper.appendChild(this.viewport);
    this.element.appendChild(this.wrapper);
    this.render(options || {});
    this.initialized = true;
  };

  // Render viewport
  Dyframe.prototype.render = function (options) {
    if (typeof options === 'object') {
      this.updateOptions(options);
    }
    if (this.waiting) {
      this.queued = true;
      return;
    }
    this.renderDom();
  };

  // Actually update DOM
  Dyframe.prototype.renderDom = function () {
    var self = this;
    this.updateClass();
    var innerSize = getInnerSize(this.element);
    this.width = innerSize.width;
    this.height = innerSize.height;
    this.wrapper.style.paddingBottom = this.height + 'px';
    this.scale();
    this.viewport.contentWindow.document.open();
    this.viewport.contentWindow.document.write(this.options.html);
    this.viewport.contentWindow.document.close();
    this.queued = false;
    if (this.options.interval > 0) {
      this.waiting = true;
      setTimeout(function () {
        self.waiting = false;
        if (self.queued) {
          self.renderDom();
        }
      }, this.options.interval);
    }
  };

  // Init or override options
  Dyframe.prototype.updateOptions = function (options) {
    if (!this.options) {
      this.options = mergeObjects({}, defaults, options);
      return;
    }
    mergeObjects(this.options, options);
  };

  // Check if active profile is given
  Dyframe.prototype.hasActiveProfile = function () {
    return !!(this.options.profile && profiles[this.options.profile]);
  };

  // Update class name of dyframe.element
  Dyframe.prototype.updateClass = function () {
    removePrefixedClass(this.element, prefix + 'profile-');
    if (this.hasActiveProfile()) {
      addClass(this.element, prefix + 'profile-' + this.options.profile);
    }
  };

  // Scale preview accroding to options
  Dyframe.prototype.scale = function () {
    var scale = this.width / this.getViewportWidth();
    setStyles(this.viewport, {
      width: (100 / scale) + '%',
      height: (100 / scale) + '%',
      webkitTransform: 'scale(' + scale + ')',
      msTransform: 'scale(' + scale + ')',
      transform: 'scale(' + scale + ')'
    });
  };

  // Get width of rendering HTML
  Dyframe.prototype.getViewportWidth = function () {
    var config = this.hasActiveProfile() ? profiles[this.options.profile] : {
      width: this.options.width,
      deviceWidth: this.options.deviceWidth
    };
    if (!config.deviceWidth) {
      return config.width;
    }
    var viewportData = this.getViewportData();
    var width = viewportData.width;
    var initialScale = viewportData['initial-scale'];
    if (width) {
      if (width === 'device-width') {
        return config.deviceWidth;
      }
      return parseInt(width, 10);
    }
    if (initialScale && initialScale > 0) {
      return Math.floor(config.deviceWidth / parseFloat(initialScale));
    }
    return config.width;
  };

  // Get viewport content as object
  Dyframe.prototype.getViewportData = function () {
    var el = document.createElement('div');
    var viewportElement;
    var viewportContent;
    var viewportData = {};
    el.innerHTML = this.options.html;
    viewportElement = el.querySelector('meta[name="viewport"]');
    if (!viewportElement) {
      return viewportData;
    }
    viewportContent = viewportElement.getAttribute('content');
    if (!viewportContent) {
      return viewportData;
    }
    viewportContent.split(',').forEach(function (configSet) {
      var config = configSet.trim().split('=');
      if (!config[0] || !config[1]) {
        return;
      }
      viewportData[config[0].trim()] = config[1].trim();
    });
    return viewportData;
  };

  // Clean up element and remove classes
  Dyframe.prototype.destroy = function () {
    if (!this.initialized) {
      return;
    }
    removeClass(this.element, 'df-element');
    removePrefixedClass(this.element, 'df-profile-');
    this.element.removeChild(this.wrapper);
    this.initialized = false;
  };

  // Add custom profile
  Dyframe.addProfile = function (name, profileData) {
    var profileDefaults = {
      width: defaults.width,
      deviceWidth: defaults.deviceWidth
    };
    var profile = mergeObjects({}, profileDefaults, profileData);
    profiles[name] = profile;
  };

  // Utility for merging objects
  var mergeObjects = function () {
    var merged = arguments[0];
    var i;
    var prop;
    for (i = 1; i < arguments.length; i++) {
      for (prop in arguments[i]) {
        if (arguments[i].hasOwnProperty(prop)) {
          merged[prop] = arguments[i][prop];
        }
      }
    }
    return merged;
  };

  // Utility for setting styles
  var setStyles = function (element, styles) {
    var prop;
    for (prop in styles) {
      element.style[prop] = styles[prop];
    }
  };

  // Utility for adding class
  var addClass = function (element, className) {
    if (element.classList) {
      element.classList.add(className);
    }
    else {
      element.className += ' ' + className;
    }
  };

  // Utility for removing class
  var removeClass = function (element, className) {
    if (element.classList) {
      element.classList.remove(className);
    }
    else {
      var pattern = new RegExp('(^|\\s)' + className + '(?!\\S)', 'g');
      element.className = element.className.replace(pattern, '');
    }
  };

  // Utility for removing prefixed classes (e.g. "df-profile-*")
  var removePrefixedClass = function (element, classPrefix) {
    var pattern = new RegExp('(^|\\s)' + classPrefix + '\\S+', 'g');
    element.className = element.className.replace(pattern, '');
  };

  // Get inner width/height of element
  var getInnerSize = function (element) {
    var style = window.getComputedStyle(element);
    var paddingTop = parseInt(style.getPropertyValue('padding-top'), 0);
    var paddingLeft = parseInt(style.getPropertyValue('padding-left'), 0);
    var paddingRight = parseInt(style.getPropertyValue('padding-right'), 0);
    var paddingBottom = parseInt(style.getPropertyValue('padding-bottom'), 0);
    var innerSize = {
      width: element.clientWidth - (paddingLeft + paddingRight),
      height: element.clientHeight - (paddingTop + paddingBottom)
    };
    return innerSize;
  };

  return Dyframe;

}));
